posts/2023-06-29-linux-arm64-github-actions/index.qmd

```diff
@@ -0,0 +1,316 @@
+---
+title: "Testing Packages on Linux ARM64 with GitHub Actions"
+description: |
+  How to use GitHub Actions to systematically build and test a Bioconductor package on 
+  Linux ARM64 archictecture.
+author:
+  name: "Mike L. Smith"
+  url: https://github.com/grimbough
+categories: tech-notes
+---
+
+![Bioconductor on Linux ARM64](../2023-06-09-debug-linux-arm64-on-docker/bioconductor_logo_cmyk.png)
+
+# Introduction
+
+The Bioconductor Build System (BBS) now includes routine package testing on Linux
+ARM64, but the relatively low frequency of testing means this, if a problem
+occurs with your package, it can take a while to identify and fix the issue
+using the build system alone.
+
+The previous blog post "[Emulated build and test of Bioconductor packages for
+Linux
+ARM64](https://bioconductor.github.io/biocblog/posts/2023-06-09-debug-linux-arm64-on-docker/)"
+described how one can use Docker and architecture emulation to build, test, and
+debug a Bioconductor package for running on Linux ARM64 architecture when you
+only have access to local x86 hardware.
+
+However such manual testing can be frustrating to run as a package developer,
+either because it's an extra task you have to run frequently, or because you
+only do it occasionally and forget the steps involved.  Ideally such testing
+would happen automatically whenever you make changes to a package, but providing
+more rapid feedback than the BBS provides.  In this article we build on these
+previously presented ideas to describe one approach for testing package on Linux
+ARM64 using a continuous integration environment on GitHub Actions.
+
+# Workflow Implementation
+
+An example workflow implementation can be found at
+[https://github.com/grimbough/bioc-testing-with-arm64/blob/main/.github/workflows/test-package-arm64.yml](https://github.com/grimbough/bioc-testing-with-arm64/blob/main/.github/workflows/test-package-arm64.yml).
+In the remainder of this post we'll discuss how some of the implementation
+choices made there and how they work.
+
+## Choice of Docker container
+
+The first thing to remember when using architecture emulation is that everything
+works much slower than when running natively - typically by at least an order of
+magnitude. This influences some of the decisions made during this workflow
+regarding which containers to use and what we want to cache between workflow
+steps.  Some operations that might be acceptable in a standard workflow become
+painfully slow under emulation, and so we try reduce the number of slow steps.
```

Comment:
There are few open issues at Github Actions runner repo about providing Linux ARM64 runners but they are not yet available! At the moment one can use Github Actions natively on Linux ARM64 only on self-hosted runner.

There are alternatives though!
E.g. at OpenHPC project we use Cirrus CI for testing on aarch64 - https://github.com/openhpc/ohpc/blob/3.x/.cirrus.yml
If you like I could send a PR to https://github.com/grimbough/bioc-testing-with-arm64 to compare it against Github CI!

Reply:
Yes, much like M1 Mac runners the milestone for GitHub having them available keeps moving into the future.

I think it'd be great to have information on other CI options, although the feeling I get is that GitHub actions is by far the most dominante player in the Bioconductor community.  

Since this post is alredy quite long, perhaps you could put the information about Cirrus CI into another post in a few weeks, and we can have  series of articles on the topic of ARM64.  I think if Bioc really wants to support Linux ARM64 we should provide as much information to developers as possible, since the architecture is still pretty niche and relatively hard to test on.  A steady flow articles might be more likely to catch peoples attention that one massive doc.

Reply:
Good point!
I will work on the native arm64 CI based on CirrusCI after my holidays! If all goes well I will prepare a follow-up blog article!